tests/sparsity/dynamic_sparsity_pattern_compute_mmult_2.cc

```diff
@@ -0,0 +1,101 @@
+
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2004 - 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+// check DynamicSparsityPattern::compute_mmult_pattern(). Test if multiplication
+// of two pattern yield the right DynamicSparsityPattern.
+
+#include "../tests.h"
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/sparsity_pattern.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/sparse_matrix.h>
+
+
+void test ()
+{
+  // create two different FullMatrix objects and add entries randomly to it.
+  // Multiply them via FullMatrix::mmult(..), go through each row and column,
+  // and check if DynamicSparsityPattern::compute_mmult_pattern(..) predicts
+  // the entries (i,j) where values supposed to be because of the multiplication.
+
+  const unsigned int M = 100;
+  const unsigned int N = 50;
+  const unsigned int O = 75;
+
+  const unsigned int m_entries = 4;
+  const unsigned int n_entries = 5;
+  const unsigned int o_entries = 6;
+
+  bool test_failed = false;
+
+  DynamicSparsityPattern dyn_left (M,N),
+                         dyn_right (N,O);
+  DynamicSparsityPattern dyn_out;
+
+  FullMatrix<double> mat_left(M,N),
+             mat_right(N,O),
+             mat_out(M,O);
+
+  // add manually entries to pattern and FullMatrix
+  for (unsigned int i=0; i<M; i+=m_entries)
+    for (unsigned int j=0; j<N; j+=n_entries)
+      {
+        dyn_left.add (i,j);
+        mat_left[i][j] = 1;
+      }
+  // add manually entries to pattern and FullMatrix
+  for (unsigned int i=0; i<N; i+=n_entries)
+    for (unsigned int j=0; j<O; j+=o_entries)
+      {
+        dyn_right.add (i,j);
+        mat_right[i][j] = 1;
+      }
+  // add also entries into non-matching columns of *_left
```

Comment:
sorry, i don't understand how this test is different from the first one.

What I was asking is that you have some row `I: [xx000x0000]^T` in left pattern and some column `J: [000x0000xx]` so that the resulting sparsity pattern does not have entry `(I,J)`.

This loop simply adds some additional non-zeroes to the left pattern, but I don't see how it is related to the case above.

Reply:
No I understand what you are asking for!
I will realize the following two changes right away:
- where entries are and are'nt will be selected randomly
- one entry in the resulting matrix is zero because the line from the left does not match any entries of the column from the right